src/rust/engine/serverset/src/lib.rs

```diff
@@ -0,0 +1,373 @@
+  fn clone(&self) -> Self {
+    Serverset {
+      inner: self.inner.clone(),
+    }
+  }
+}
+
+struct Inner<T> {
+  servers: Vec<Backend<T>>,
+
+  // Only visible for testing
+  pub(crate) next: AtomicUsize,
+
+  failure_initial_lame: Duration,
+
+  failure_backoff_ratio: num_rational::Ratio<u32>,
+
+  failure_max_lame: Duration,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub enum Health {
+  Healthy,
+  Unhealthy,
+}
+
+#[derive(Debug)]
+struct UnhealthyInfo {
+  unhealthy_since: Instant,
+  next_attempt_after: Duration,
+}
+
+#[derive(Debug)]
+struct Backend<T> {
+  server: T,
+  unhealthy_info: Arc<Mutex<Option<UnhealthyInfo>>>,
+}
+
+#[derive(Clone, Copy, Debug)]
+pub struct BackoffConfig {
+  ///
+  /// The time a backend will be skipped after it is first reported unhealthy.
+  ///
+  pub initial_lame: Duration,
+  ///
+  /// Ratio by which to multiply the most recent lame duration if a backend continues to be
+  /// unhealthy.
+  ///
+  /// The inverse is used when easing back in after health recovery.
+  pub backoff_ratio: f64,
+
+  ///
+  /// Maximum duration to wait between attempts.
+  ///
+  pub max_lame: Duration,
+}
+
+impl<T: Clone + Send + Sync + 'static> Serverset<T> {
+  pub fn new(servers: Vec<T>, backoff_config: BackoffConfig) -> Result<Self, String> {
+    if servers.is_empty() {
+      return Err("Must supply some servers".to_owned());
+    }
+
+    let BackoffConfig {
+      initial_lame,
+      backoff_ratio,
+      max_lame,
+    } = backoff_config;
+
+    if backoff_ratio < 1.0 {
+      return Err(format!(
+        "Failure backoff ratio must be at least 1, got: {}",
+        backoff_ratio
+      ));
+    }
+    let backoff_ratio =
+      num_rational::Ratio::<i8>::approximate_float(backoff_ratio).ok_or_else(|| {
+        format!(
+          "Couldn't find reasonable backoff ratio for {}",
+          backoff_ratio
+        )
+      })?;
+    let backoff_ratio =
+      num_rational::Ratio::new(*backoff_ratio.numer() as u32, *backoff_ratio.denom() as u32);
+    Ok(Serverset {
+      inner: Arc::new(Inner {
+        servers: servers
+          .into_iter()
+          .map(|s| Backend {
+            server: s,
+            unhealthy_info: Arc::new(Mutex::new(None)),
+          }).collect(),
+        next: AtomicUsize::new(0),
+        failure_initial_lame: initial_lame,
+        failure_backoff_ratio: backoff_ratio,
+        failure_max_lame: max_lame,
+      }),
+    })
+  }
+
+  ///
+  /// Get the next (probably) healthy backend to use.
+  ///
+  /// The caller will be given a backend to use, and should call the supplied callback with the
+  /// observed health of that backend.
+  ///
+  /// If the callback is not called, the health status of the server will not be changed from its
+  /// last known status.
+  ///
+  /// If all resources are unhealthy, this function will block the calling thread until the backoff
+  /// period has completed. We'd probably prefer to use some Future-based scheduling, but that
```

Comment:
Maybe add a TODO for the Future? (a little, unfunny pun intended)

Reply:
Done